module/zfs/dmu_send.c

```diff
@@ -663,8 +664,6 @@ dump_freeobjects(dmu_send_cookie_t *dscp, uint64_t firstobj, uint64_t numobjs)
 			return (SET_ERROR(EINTR));
 		dscp->dsc_pending_op = PENDING_NONE;
 	}
-	if (numobjs == 0)
-		numobjs = UINT64_MAX - firstobj;
```

Comment:
So this was only hit in the corner case mentioned above?  (where firstobj==maxobj)  That seems to be inconsistent with the bug report which had `firstobj = 37 numobjs = 18446744073709551578`, because  maxobj should be a multiple of 32.  So in the bug report, how did we get `numobjs=0` and then change it to a huge value?

Reply:
I pulled the tip of master and instrumented dump_freeobjects.  There are 3 outputs 1) on function entry 2) after the limiting code 3) after the if (numobjs ==0) numobjs = UINT64_MAX - firstobj.

NOTICE: (1) firstobj 64 numobjs 32 maxobj 64
NOTICE: (2) firstobj 64 numobjs 0 maxobj 64
NOTICE: (3) firstobj 64 numobjs 18446744073709551551 maxobj 64
NOTICE: (1) firstobj 96 numobjs 32 maxobj 64

It turns out that at 64, we get a numobjs of 0 and the UINT64_MAX code kicks in.  Also this function starts with firstobj 8 (one more than the last inode in the dataset), and counts up to 64, then goes for quite some ways by increments of 32.

Reply:
```
NOTICE: (1) firstobj 8 numobjs 1 maxobj 64
NOTICE: (2) firstobj 8 numobjs 1 maxobj 64
NOTICE: (3) firstobj 8 numobjs 1 maxobj 64
NOTICE: (1) firstobj 9 numobjs 1 maxobj 64
NOTICE: (2) firstobj 9 numobjs 1 maxobj 64
NOTICE: (3) firstobj 9 numobjs 1 maxobj 64
NOTICE: (1) firstobj 10 numobjs 1 maxobj 64
NOTICE: (2) firstobj 10 numobjs 1 maxobj 64
NOTICE: (3) firstobj 10 numobjs 1 maxobj 64
NOTICE: (1) firstobj 11 numobjs 1 maxobj 64
NOTICE: (2) firstobj 11 numobjs 1 maxobj 64
NOTICE: (3) firstobj 11 numobjs 1 maxobj 64
NOTICE: (1) firstobj 12 numobjs 1 maxobj 64
NOTICE: (2) firstobj 12 numobjs 1 maxobj 64
NOTICE: (3) firstobj 12 numobjs 1 maxobj 64
NOTICE: (1) firstobj 13 numobjs 1 maxobj 64
NOTICE: (2) firstobj 13 numobjs 1 maxobj 64
NOTICE: (3) firstobj 13 numobjs 1 maxobj 64
NOTICE: (1) firstobj 14 numobjs 1 maxobj 64
NOTICE: (2) firstobj 14 numobjs 1 maxobj 64
NOTICE: (3) firstobj 14 numobjs 1 maxobj 64
NOTICE: (1) firstobj 15 numobjs 1 maxobj 64
NOTICE: (2) firstobj 15 numobjs 1 maxobj 64
NOTICE: (3) firstobj 15 numobjs 1 maxobj 64
NOTICE: (1) firstobj 16 numobjs 1 maxobj 64
NOTICE: (2) firstobj 16 numobjs 1 maxobj 64
NOTICE: (3) firstobj 16 numobjs 1 maxobj 64
NOTICE: (1) firstobj 17 numobjs 1 maxobj 64
NOTICE: (2) firstobj 17 numobjs 1 maxobj 64
NOTICE: (3) firstobj 17 numobjs 1 maxobj 64
NOTICE: (1) firstobj 18 numobjs 1 maxobj 64
NOTICE: (2) firstobj 18 numobjs 1 maxobj 64
NOTICE: (3) firstobj 18 numobjs 1 maxobj 64
NOTICE: (1) firstobj 19 numobjs 1 maxobj 64
NOTICE: (2) firstobj 19 numobjs 1 maxobj 64
NOTICE: (3) firstobj 19 numobjs 1 maxobj 64
NOTICE: (1) firstobj 20 numobjs 1 maxobj 64
NOTICE: (2) firstobj 20 numobjs 1 maxobj 64
NOTICE: (3) firstobj 20 numobjs 1 maxobj 64
NOTICE: (1) firstobj 21 numobjs 1 maxobj 64
NOTICE: (2) firstobj 21 numobjs 1 maxobj 64
NOTICE: (3) firstobj 21 numobjs 1 maxobj 64
NOTICE: (1) firstobj 22 numobjs 1 maxobj 64
NOTICE: (2) firstobj 22 numobjs 1 maxobj 64
NOTICE: (3) firstobj 22 numobjs 1 maxobj 64
NOTICE: (1) firstobj 23 numobjs 1 maxobj 64
NOTICE: (2) firstobj 23 numobjs 1 maxobj 64
NOTICE: (3) firstobj 23 numobjs 1 maxobj 64
NOTICE: (1) firstobj 24 numobjs 1 maxobj 64
NOTICE: (2) firstobj 24 numobjs 1 maxobj 64
NOTICE: (3) firstobj 24 numobjs 1 maxobj 64
NOTICE: (1) firstobj 25 numobjs 1 maxobj 64
NOTICE: (2) firstobj 25 numobjs 1 maxobj 64
NOTICE: (3) firstobj 25 numobjs 1 maxobj 64
NOTICE: (1) firstobj 26 numobjs 1 maxobj 64
NOTICE: (2) firstobj 26 numobjs 1 maxobj 64
NOTICE: (3) firstobj 26 numobjs 1 maxobj 64
NOTICE: (1) firstobj 27 numobjs 1 maxobj 64
NOTICE: (2) firstobj 27 numobjs 1 maxobj 64
NOTICE: (3) firstobj 27 numobjs 1 maxobj 64
NOTICE: (1) firstobj 28 numobjs 1 maxobj 64
NOTICE: (2) firstobj 28 numobjs 1 maxobj 64
NOTICE: (3) firstobj 28 numobjs 1 maxobj 64
NOTICE: (1) firstobj 29 numobjs 1 maxobj 64
NOTICE: (2) firstobj 29 numobjs 1 maxobj 64
NOTICE: (3) firstobj 29 numobjs 1 maxobj 64
NOTICE: (1) firstobj 30 numobjs 1 maxobj 64
NOTICE: (2) firstobj 30 numobjs 1 maxobj 64
NOTICE: (3) firstobj 30 numobjs 1 maxobj 64
NOTICE: (1) firstobj 31 numobjs 1 maxobj 64
NOTICE: (2) firstobj 31 numobjs 1 maxobj 64
NOTICE: (3) firstobj 31 numobjs 1 maxobj 64
NOTICE: (1) firstobj 37 numobjs 1 maxobj 64
NOTICE: (2) firstobj 37 numobjs 1 maxobj 64
NOTICE: (3) firstobj 37 numobjs 1 maxobj 64
NOTICE: (1) firstobj 38 numobjs 1 maxobj 64
NOTICE: (2) firstobj 38 numobjs 1 maxobj 64
NOTICE: (3) firstobj 38 numobjs 1 maxobj 64
NOTICE: (1) firstobj 39 numobjs 1 maxobj 64
NOTICE: (2) firstobj 39 numobjs 1 maxobj 64
NOTICE: (3) firstobj 39 numobjs 1 maxobj 64
NOTICE: (1) firstobj 40 numobjs 1 maxobj 64
NOTICE: (2) firstobj 40 numobjs 1 maxobj 64
NOTICE: (3) firstobj 40 numobjs 1 maxobj 64
NOTICE: (1) firstobj 41 numobjs 1 maxobj 64
NOTICE: (2) firstobj 41 numobjs 1 maxobj 64
NOTICE: (3) firstobj 41 numobjs 1 maxobj 64
NOTICE: (1) firstobj 42 numobjs 1 maxobj 64
NOTICE: (2) firstobj 42 numobjs 1 maxobj 64
NOTICE: (3) firstobj 42 numobjs 1 maxobj 64
NOTICE: (1) firstobj 43 numobjs 1 maxobj 64
NOTICE: (2) firstobj 43 numobjs 1 maxobj 64
NOTICE: (3) firstobj 43 numobjs 1 maxobj 64
NOTICE: (1) firstobj 44 numobjs 1 maxobj 64
NOTICE: (2) firstobj 44 numobjs 1 maxobj 64
NOTICE: (3) firstobj 44 numobjs 1 maxobj 64
NOTICE: (1) firstobj 45 numobjs 1 maxobj 64
NOTICE: (2) firstobj 45 numobjs 1 maxobj 64
NOTICE: (3) firstobj 45 numobjs 1 maxobj 64
NOTICE: (1) firstobj 46 numobjs 1 maxobj 64
NOTICE: (2) firstobj 46 numobjs 1 maxobj 64
NOTICE: (3) firstobj 46 numobjs 1 maxobj 64
NOTICE: (1) firstobj 47 numobjs 1 maxobj 64
NOTICE: (2) firstobj 47 numobjs 1 maxobj 64
NOTICE: (3) firstobj 47 numobjs 1 maxobj 64
NOTICE: (1) firstobj 48 numobjs 1 maxobj 64
NOTICE: (2) firstobj 48 numobjs 1 maxobj 64
NOTICE: (3) firstobj 48 numobjs 1 maxobj 64
NOTICE: (1) firstobj 49 numobjs 1 maxobj 64
NOTICE: (2) firstobj 49 numobjs 1 maxobj 64
NOTICE: (3) firstobj 49 numobjs 1 maxobj 64
NOTICE: (1) firstobj 50 numobjs 1 maxobj 64
NOTICE: (2) firstobj 50 numobjs 1 maxobj 64
NOTICE: (3) firstobj 50 numobjs 1 maxobj 64
NOTICE: (1) firstobj 51 numobjs 1 maxobj 64
NOTICE: (2) firstobj 51 numobjs 1 maxobj 64
NOTICE: (3) firstobj 51 numobjs 1 maxobj 64
NOTICE: (1) firstobj 52 numobjs 1 maxobj 64
NOTICE: (2) firstobj 52 numobjs 1 maxobj 64
NOTICE: (3) firstobj 52 numobjs 1 maxobj 64
NOTICE: (1) firstobj 53 numobjs 1 maxobj 64
NOTICE: (2) firstobj 53 numobjs 1 maxobj 64
NOTICE: (3) firstobj 53 numobjs 1 maxobj 64
NOTICE: (1) firstobj 54 numobjs 1 maxobj 64
NOTICE: (2) firstobj 54 numobjs 1 maxobj 64
NOTICE: (3) firstobj 54 numobjs 1 maxobj 64
NOTICE: (1) firstobj 55 numobjs 1 maxobj 64
NOTICE: (2) firstobj 55 numobjs 1 maxobj 64
NOTICE: (3) firstobj 55 numobjs 1 maxobj 64
NOTICE: (1) firstobj 56 numobjs 1 maxobj 64
NOTICE: (2) firstobj 56 numobjs 1 maxobj 64
NOTICE: (3) firstobj 56 numobjs 1 maxobj 64
NOTICE: (1) firstobj 57 numobjs 1 maxobj 64
NOTICE: (2) firstobj 57 numobjs 1 maxobj 64
NOTICE: (3) firstobj 57 numobjs 1 maxobj 64
NOTICE: (1) firstobj 58 numobjs 1 maxobj 64
NOTICE: (2) firstobj 58 numobjs 1 maxobj 64
NOTICE: (3) firstobj 58 numobjs 1 maxobj 64
NOTICE: (1) firstobj 59 numobjs 1 maxobj 64
NOTICE: (2) firstobj 59 numobjs 1 maxobj 64
NOTICE: (3) firstobj 59 numobjs 1 maxobj 64
NOTICE: (1) firstobj 60 numobjs 1 maxobj 64
NOTICE: (2) firstobj 60 numobjs 1 maxobj 64
NOTICE: (3) firstobj 60 numobjs 1 maxobj 64
NOTICE: (1) firstobj 61 numobjs 1 maxobj 64
NOTICE: (2) firstobj 61 numobjs 1 maxobj 64
NOTICE: (3) firstobj 61 numobjs 1 maxobj 64
NOTICE: (1) firstobj 62 numobjs 1 maxobj 64
NOTICE: (2) firstobj 62 numobjs 1 maxobj 64
NOTICE: (3) firstobj 62 numobjs 1 maxobj 64
NOTICE: (1) firstobj 63 numobjs 1 maxobj 64
NOTICE: (2) firstobj 63 numobjs 1 maxobj 64
NOTICE: (3) firstobj 63 numobjs 1 maxobj 64
NOTICE: (1) firstobj 64 numobjs 32 maxobj 64
NOTICE: (2) firstobj 64 numobjs 0 maxobj 64
NOTICE: (3) firstobj 64 numobjs 18446744073709551551 maxobj 64
NOTICE: (1) firstobj 96 numobjs 32 maxobj 64
NOTICE: (1) firstobj 128 numobjs 32 maxobj 64
...
...
NOTICE: (1) firstobj 32672 numobjs 32 maxobj 64
NOTICE: (1) firstobj 32704 numobjs 32 maxobj 64
NOTICE: (1) firstobj 32736 numobjs 32 maxobj 64
NOTICE: (1) firstobj 32768 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 65536 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 98304 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 131072 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 163840 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 196608 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 229376 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 262144 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 294912 numobjs 32768 maxobj 64
...
...
NOTICE: (1) firstobj 33423360 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 33456128 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 33488896 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 33521664 numobjs 32768 maxobj 64
NOTICE: (1) firstobj 33554432 numobjs 33554432 maxobj 64
NOTICE: (1) firstobj 67108864 numobjs 33554432 maxobj 64
NOTICE: (1) firstobj 100663296 numobjs 33554432 maxobj 64
NOTICE: (1) firstobj 134217728 numobjs 33554432 maxobj 64
NOTICE: (1) firstobj 167772160 numobjs 33554432 maxobj 64
...
...
NOTICE: (1) firstobj 893353197568 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 927712935936 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 962072674304 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 996432412672 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 1030792151040 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 1065151889408 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 1099511627776 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 1133871366144 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 1168231104512 numobjs 34359738368 maxobj 64
...
...
NOTICE: (1) firstobj 35046933135360 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 35081292873728 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 35115652612096 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 35150012350464 numobjs 34359738368 maxobj 64
NOTICE: (1) firstobj 35184372088832 numobjs 35184372088832 maxobj 64
NOTICE: (1) firstobj 70368744177664 numobjs 35184372088832 maxobj 64
NOTICE: (1) firstobj 105553116266496 numobjs 35184372088832 maxobj 64
NOTICE: (1) firstobj 140737488355328 numobjs 35184372088832 maxobj 64
NOTICE: (1) firstobj 175921860444160 numobjs 35184372088832 maxobj 64
...
...
NOTICE: (1) firstobj 35923243902697472 numobjs 35184372088832 maxobj 64
NOTICE: (1) firstobj 35958428274786304 numobjs 35184372088832 maxobj 64
NOTICE: (1) firstobj 35993612646875136 numobjs 35184372088832 maxobj 64
NOTICE: (1) firstobj 36028797018963968 numobjs 36028797018963968 maxobj 64
NOTICE: (1) firstobj 72057594037927936 numobjs 36028797018963968 maxobj 64
```

Reply:
FWIW I changed only the off by one error as Matt suggested and that seems to work.  Then I don't know if we need the if (numobjs ==0) numobjs = UINT64_MAX - firstobj code anymore.